runtime/src/bank.rs

```diff
@@ -871,6 +877,11 @@ impl Bank {
     pub fn get_epoch_and_slot_index(&self, slot: u64) -> (u64, u64) {
         self.epoch_schedule.get_epoch_and_slot_index(slot)
     }
+
+    pub fn is_votable(&self) -> bool {
+        let max_tick_height = (self.slot + 1) * self.ticks_per_slot - 1;
+        self.is_delta.load(Ordering::Relaxed) && self.tick_height() == max_tick_height
```

Comment:
This pub function (called from another crate) needs tests and a doc comment. At first glance, it looks like this implementation can be replaced with:

```
pub is_votable(&self) -> bool {
   self.is_frozen() && !self.accounts.is_empty()
}
```

but no way to find out from within the solana_runtime crate.  The test up in ReplayStage only tests one case, but this function alone has several edge cases.

Reply:
Those are not equivalent, is_delta can be false when !self.accounts.is_empty() is true (f your parent bank is nonempty, the child bank will always be nonempty as well, even if the child processes no transactions). We check self.tick_height() == max_tick_height instead of is_frozen because @rob-solana pointed out we may want to check if a bank is votable before freezing it.

Reply:
Also, which edge cases were you thinking of. The one I can think of is that a bank that processes only ticks has self.is_delta =- false

Reply:
@carllin there definitely should be local tests.  cargo test --manifest-path runtime/Cargo.toml should cover everything in here

@garious this check is conservative and so shouldn't have edge cases.  once tick_height() is max, no commit() will be called, and commit() is where we move to "is_delta".

the reason this check is down here is because otherwise, its logic would have to be duplicated in the TVU and in the TPU (which talk to the bank at different levels)

Reply:
Why not use HashQueue::hash_height()? I don't really want an answer as much as I want a test to answer on your behalf.